compiler/rustc_ast_pretty/src/pprust/state.rs

```diff
@@ -1411,6 +1415,12 @@ impl<'a> State<'a> {
             self.maybe_print_comment(field.span.lo());
             self.print_outer_attributes(&field.attrs);
             self.print_visibility(&field.vis);
+            if matches!(
+                &field.ty.kind,
+                ast::TyKind::AnonymousStruct(..) | ast::TyKind::AnonymousUnion(..)
+            ) {
+                self.head("");
+            }
```

Comment:
Could you add a test for
```rust
type A = struct { field: u8 };
```
?

Printing for `TyKind::Anonymous(Struct,Union)` is supposed to be self-contained and not relying on `_` field preparing something before it. I'm not sure whether unbalanced `end()` can cause issues.

Reply:
Maybe passing a boolean indicating whether the struct should call `self.head("struct")` or `self.s.word("struct")`?

In that case I would rename `print_ty` to `print_ty_maybe_head` and make a new `print_ty` calling `print_ty_maybe_head(ty, true)`, and only using `print_ty_maybe_head(ty, false)` on `print_record_struct_body`. That way the api doesn't change and we correctly preserve the formatting as intended.

Reply:
the other option is to return to the hardcoded `cbox`/`ibox` and play with the indents until everything works.

Reply:
Another option would be to directly print the `AnonymousStruct`s on `print_record_struct_body` and using a `head` on `print_ty`.

Reply:
I think you are overthinking, this is not `rustfmt`, the goal here is to 1) avoid ICE and 2) make the logic as simple as possible.
I wouldn't want to spend any more time on tweaking boxes, the initial version of this PR was good enough.